### ISSUE TITLE
[E602] Don't compare to empty string - what's a better way?

### ISSUE THREAD

**Issue by geerlingguy** [ansible-community/ansible-lint#457]
# Issue Type

- Bug report

# Ansible and Ansible Lint details

```
$ ansible --version
ansible 2.7.5
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/Users/jgeerling/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.15 (default, Jul 17 2018, 19:41:20) [GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.2)]

$ ansible-lint --version
ansible-lint 4.0.0
```

- ansible installation method: pip
- ansible-lint installation method: pip

# Desired Behaviour

Using a string comparison to check for a string being empty should be A-OK:

```yaml
- name: Do something with my_var.
  command: "do_something_with {{ my_var }}"
  when: my_var != ''
```

ansible-lint should throw no errors for this code.

# Actual Behaviour (Bug report only)

ansible-lint throws the following:

```
[602] Don't compare to empty string
```

The problem is, if I set it to:

```yaml
- name: Do something with my_var.
  command: "do_something_with {{ my_var }}"
  when: my_var
```

Then I get the error:

```
fatal: [localhost]: FAILED! =>
The conditional check 'my_var' failed. The error was: error while evaluating conditional (my_var): 'testing' is undefined
```

So I'm wondering—if we're not supposed to compare to empty strings in a duck-typed language like Python, what _are_ we supposed to do? There are a ton of use cases for comparison to an empty string.

## Workarounds

Currently there are two ways to get past this:

  1. Head-in-the-sand:
     1. Add `.ansible-lint` file in your role/playbook.
     1. Add the following to skip the rule entirely:

            skip_list:
              - '602'

  2. Bypass the rule's regex:

         when: my_var | length > 0

Number 1 is just ignoring the issue. Number 2 is more obtuse and makes for more unreadable code (at least IMO)... but is that preferred if we truly need to check if a string has a value?

**Comment by geerlingguy** [comment]
For reference, original discussion was from: https://github.com/ansible/ansible-lint/issues/450

**Comment by geerlingguy** [comment]
So another workaround could be to use the `bool` filter (see https://docs.ansible.com/ansible/latest/user_guide/playbooks_filters.html):

```
when: my_var | bool
```

Is that the preferred way rather than a string comparison? I still don't think it's important / best-practice-y enough to force one way over the other, but that's a better compromise than the length filter at least—it makes more semantic sense.

**Comment by geerlingguy** [comment]
Hmm... but with the `bool` filter, it looks like _there be dragons_: https://github.com/ansible/ansible/issues/25893

**Comment by Stolb27** [comment]
3. use not flat variable name in when expression, like vars.my_var (magic var vars)

**Comment by drybjed** [comment]
How about providing a "default wrapper" for the variable?
```yaml
when: my_var|d()
```
That should behave as expected during Ansible execution and not trigger the E602 rule.

**Comment by sseide** [comment]
I for myself do not understand the rationale behind this rule...

Using the same example as Jeff Geerling above with `my_var`

1) my_var contains a string - following this linter rule Ansible playbook breaks and does not work at all.
2) my_var contains a boolean or number  - why on earth should someone compare that again an empty string?
3) my_var contains a string and i have to use some of the - @geerlingguy and others, please do not take any offense - use some of the ugly work around above is neither elegant, comprehensible nor easy maintainable. 

IMHO this rule should just be dropped. Or at least not flagged as error importing roles into Ansible Galaxy.

And finally - up to now no member of the Ansible team has come up with a explanation for this rule and the things it tries to solve...

**Comment by willthames** [comment]
Using `when: my_var` seems reasonable to me, and it's not clear why ansible then tries to evaluate the contents of `my_var` - seems like some double evaluation is going on.

Having said that, this code hasn't worked since 2.1 and I don't have an older version to test with, so I'd be tempted to drop this rule, it certainly doesn't work as I originally envisaged it (my guess is that it might have worked in 1.9, presumably I had a working alternative at some point)

This also illustrates that some rules are missing tests!

**Comment by tgadiev** [comment]
How about making `when: my_var` conditional working in Ansible for string variables? It would be good alternative to empty string comparison.

**Comment by sseide** [comment]
When Ansible would be changed to work that way (`when: my_var`), i would be happy to follow this rule...
IMHO not working with strings right now is the only big reason against this one...

**Comment by tgadiev** [comment]
Looks like it works now but only for hash variables. Using simple variables in conditionals causes trouble `when: my_var`, but hash variables in conditionals works as expected: `when: my_hash.my_var`. See 
[ansible/ansible#39414](https://github.com/ansible/ansible/issues/39414) for related details. 

So, the good workaround for now is avoid using simple variables in string conditionals like `when: my_var`. If possible they should be changed to hashed ones. Several related variables can be grouped in one hash for example.

**Comment by geerlingguy** [comment]
@tgadiev - That workaround doesn't seem very helpful for this use case though; I often try to split my variables into simpler individual variables so if I want to override one in a playbook I don't have to end up overriding an entire hash/object with 3, 5, 10, or dozens of variables.

**Comment by tgadiev** [comment]
@geerlingguy Are you talking about overriding single key-value item of an entire hash? It's quite simple using `combine` filter. You don't have to override an entire hash/object to change single parameter.

**Comment by geerlingguy** [comment]
That particular bug report's been open for almost a year, and the bug has _existed_ for at least a couple years, so yes, I believe this rule can and should be dropped unless that bug is actually fixed.

On a separate level (but again, unrelated to the discussion here), I don't particularly like duck typing, so explicitly comparing to a string is something I often do in any non-strict language (like PHP as well), and I'm not sure why there's a rule saying that is a bad thing...

**Comment by ViachaslauKabak** [comment]
@geerlingguy , even in PHP there is a method to check if string is empty
https://www.rosettacode.org/wiki/Empty_string#PHP
let's use the better code style)

**Comment by tgadiev** [comment]
@geerlingguy, 
Ansible came from Python world. It's supposed to follow [Zen of Python](https://en.wikipedia.org/wiki/Zen_of_Python) principles and related codestyle conventions. As for conditionals, Python code guides recommends to use direct truth value testing. See [these recommendations](https://www.python.org/dev/peps/pep-0008/#programming-recommendations) for example. Short quote:
> For sequences, (strings, lists, tuples), use the fact that empty sequences are false.
>```
>Yes: if not seq:
>      if seq:
>
>No:  if len(seq):
>      if not len(seq):
>```
And below:

> Don't compare boolean values to True or False using ==.
> ```
> Yes:   if greeting:
> No:    if greeting == True:
> Worse: if greeting is True:
> ```

I guess Ansible code should also follow these recommendations. If you don't like it and prefer other program language code style - it's up to you. Ansible is quite flexible and allows various code styles. If you prefer php-like conditionals for string variables and you're happy with that approach - it's ok, your code will work. But why do you need ansible-lint for that? You can drop any rule you don't like on your own level just disabling it in config. Why do you ask to drop it on general level?

**Comment by sseide** [comment]
@tgadiev Thanks for your explanations of python code style...

Just to summaries this:
* There exists an long standing bug in Ansible ansible/ansible#39414 nobody seems to care about.
* This bug prevents us following this linter rule, which, as you said has its merrits

Most simple solution:
Ansible bug gets fixed, rule works as expected and all people are happy as there are no more workarounds needed. And if someone (for whatever reason) really does not like this rule it can be disabled...

Second solution:
Bug is ignored further and this rule creates lot of extra work for people completly destroying possible atempts to follow the zen of python... Than this rule should be dropped for beeing inconsistent.

Third solution:
Get everyone ignore the rule inside their playbooks. But whats the reason of a rule everyone ignores?

I definitely prefer the first one...

**Comment by tgadiev** [comment]
> I definitely prefer the first one...
So do I

**Comment by geerlingguy** [comment]
Ditto. I don’t care if we prefer that style, but if Ansible breaks if we follow that style and this lint rule slaps me on the wrist if I follow that style (I know I can disable it... in all my 150+ roles and 500+ playbooks individually), then I agree either that big needs to be elevated in priority or this rule is toothless.

**Comment by tgadiev** [comment]
Looks like the fixes are coming. Let's wait for related PR [#51030](https://github.com/ansible/ansible/pull/51030) to be merged.

**Comment by insideClaw** [comment]
+1, just had to fix the error with the aforementioned
`when myvar | length > 0`
and it doesn't seem like anything but dirty and convoluted when you consider that the discussed "if variable:" Pythonic approach does not work here.

**Comment by tgadiev** [comment]
Related PR [#51030](https://github.com/ansible/ansible/pull/51030) has been merged to develop.

**Comment by tgadiev** [comment]
Who wants to test?

**Comment by mihai-satmarean** [comment]
Hi, having some similar issue, maybe I can find some advice here


```
- name: nice task
  include_tasks: "nice/RedHat.yml"
  when:
    - action is defined and action == "read"
    - ansible_distribution == 'RedHat' or ansible_distribution == "CentOS"

```

results in 

> Couldn't parse task at Unknown (no action detected in task. This often indicates a misspelled module name, or incorrect module path.)


Thanks!

**Comment by odyssey4me** [comment]
Perhaps I'm misunderstanding how this is supposed to work, but using Ansible 2.8.2 (which includes the above-mentioned fix), I expected this playbook to work:

```
- hosts: localhost
  connection: local
  gather_facts: no
  vars:
    foo: ""
    bar: "baz"
  tasks:
    - debug:
        msg: "foo is empty"
      when:
        - foo == ""

    - debug:
        msg: "foo is empty"
      when:
        - not foo

    - debug:
        msg: "bar is not empty"
      when:
        - bar != ""

    - debug:
        msg: "bar is not empty"
      when:
        - bar
```

But the result is:

```
PLAY [localhost] *****************************************************************************************************

TASK [debug] *********************************************************************************************************
ok: [localhost] => {}

MSG:

foo is empty


TASK [debug] *********************************************************************************************************
ok: [localhost] => {}

MSG:

foo is empty


TASK [debug] *********************************************************************************************************
ok: [localhost] => {}

MSG:

bar is not empty


TASK [debug] *********************************************************************************************************
[DEPRECATION WARNING]: evaluating baz as a bare variable, this behaviour will go away and you might need to add |bool
 to the expression in the future. Also see CONDITIONAL_BARE_VARS configuration toggle.. This feature will be removed 
in version 2.12. Deprecation warnings can be disabled by setting deprecation_warnings=False in ansible.cfg.
fatal: [localhost]: FAILED! => {}

MSG:

The conditional check 'bar' failed. The error was: error while evaluating conditional (bar): 'baz' is undefined

The error appears to be in '/home/jpretori/code/redhat-openstack/infrared/test.yml': line 23, column 7, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:


    - debug:
      ^ here



PLAY RECAP ***********************************************************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   
```
How is this rule meant to be satisfied?

**Comment by tgadiev** [comment]
@odyssey4me Now there are two options to handle this:
1. Set CONDITIONAL_BARE_VARS param to False in config or environment
e.g.: export ANSIBLE_CONDITIONAL_BARE_VARS=False
2. Adjust your code to add boolean filter:
```yaml
- debug:
    msg: "bar is not empty"
  when:
    - bar|bool
```

**Comment by tgadiev** [comment]
I would suggest to set CONDITIONAL_BARE_VARS to false and use normal syntax from initial example.

**Comment by odyssey4me** [comment]
@tgadiev Yeah, option 2 would result in unexpected behaviour if bar happened to contain a string which matches anything considered a boolean, eg: 'No' or 'False'

**Comment by ssbarnea** [comment]
As since Ansible introduced a newer deprecation, I raised the question at https://www.reddit.com/r/ansible/comments/e31qj5/is_there_a_recommended_way_to_test_empty_values/ -- lets follow it there and see what will be the action needed. 

Until we get some kind of consensus you are better of disabling E602.

**Comment by jonpugh** [comment]
Check this out:

```
- hosts: localhost
  vars:
    empty_string: ""
    foo: ""
  tasks:
    - debug:
        msg: "foo is empty"
      when:
        - foo == empty_string
```

:laughing:

**Comment by bbaassssiiee** [comment]
> That particular bug report's been open for almost a year, and the bug has _existed_ for at least a couple years, so yes, I believe this rule can and should be dropped unless that bug is actually fixed.
> 
> On a separate level (but again, unrelated to the discussion here), I don't particularly like duck typing, so explicitly comparing to a string is something I often do in any non-strict language (like PHP as well), and I'm not sure why there's a rule saying that is a bad thing...

It has been [fixed](https://github.com/ansible/ansible/pull/51030) for almost a year, no need to be pedantic anymore @geerlingguy

**Comment by geerlingguy** [comment]
If that is fixed finally (I haven't yet confirmed it but hopefully it is), then the best way for this would be to use the Pythonic way:

```
when: gitlab_version
```

**Comment by geerlingguy** [comment]
So it is... I just tested this playbook and the tasks were skipped and run accordingly:

```yaml
---
- hosts: localhost
  connection: local
  gather_facts: false

  vars:
    string_var: ''

  tasks:
    - name: This task should be skipped.
      command: date
      when: string_var | bool

    - name: This task should not be skipped.
      command: date
      when: not string_var | bool
```

It looks like I can finally start dropping all my 'skip 602' rules in `.ansible-lint` files...

**Comment by hogarthj** [comment]
@geerlingguy it's still not quite working as expected though for the idea of testing against an empty string ... instead of an empty string use a string of `true` or `false` (or yes no etc) ... the `| bool` will use that contents to get a result, not whether the string is actually empty or not ... so if you want to test for an *empty* string you cannot do it via that syntax.

```
               
[james.hogarth@localhost hydra-perf-test]$ ansible-playbook test.yml 
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match
'all'


PLAY [localhost] *****************************************************************************************************

TASK [This task should be skipped.] **********************************************************************************
skipping: [localhost]

TASK [This task should not be skipped.] ******************************************************************************
changed: [localhost]

TASK [This string check should not be skipped.] **********************************************************************
skipping: [localhost]

TASK [This string check should not be skipped.] **********************************************************************
skipping: [localhost]

PLAY RECAP ***********************************************************************************************************
localhost                  : ok=1    changed=1    unreachable=0    failed=0    skipped=3    rescued=0    ignored=0   

[james.hogarth@localhost hydra-perf-test]$ cat test.yml
---
- hosts: localhost
  connection: local
  gather_facts: false

  vars:
    string_var: ''
    string_false: 'false'
    string_true: 'true'

  tasks:
    - name: This task should be skipped.
      command: date
      when: string_var | bool

    - name: This task should not be skipped.
      command: date
      when: not string_var | bool

    - name: This string check should not be skipped.
      command: date
      when: string_false | bool

    - name: This string check should not be skipped.
      command: date
      when: not string_true | bool
```

**Comment by geerlingguy** [comment]
Hmm... and my Jenkins role is exploding once I try switching to the Pythonic style:

```yaml
- debug: var=jenkins_repo_url

- name: Ensure Jenkins repo is installed.
  get_url:
    url: "{{ jenkins_repo_url }}"
    dest: /etc/yum.repos.d/jenkins.repo
  when: jenkins_repo_url | bool
```

Result in:

```
TASK [geerlingguy.jenkins : debug] *********************************************
ok: [instance] => {
    "jenkins_repo_url": "https://pkg.jenkins.io/redhat/jenkins.repo"
}
    
TASK [geerlingguy.jenkins : Ensure Jenkins repo is installed.] *****************
skipping: [instance]
```

Shouldn't `jenkins_repo_url | bool` result in the conditional evaluating to `True`? Am I missing something obvious here, or is Ansible still broken in its conditional evaluation?

**Comment by geerlingguy** [comment]
Ah, asking around, I discovered that the `| bool` filter only returns `True` for _truthy_ string values (so like `'true'`, or `'True'` or `1`), but not for _non-empty strings_. Big miss on my part.

In Ansible 2.10, there will be two new filters, `truthy` and `falsey` (see https://github.com/ansible/ansible/pull/62602). These filters will allow easier evaluation of duck-typed variables.

In the mean time, I think I might go with the following style instead of `| bool`, to preserve compatibility with older Ansible versions (especially considering 2.10 isn't out yet!):

```
jenkins_repo_url | default(false)
```

That filter seems to evaluate correctly to `False` if the URL is string var is empty, and doesn't give the bare conditional warning.

**Comment by ssbarnea** [comment]
As I did not receive any clear feedback from Ansible team regarding directions, I am still using the ugly `# noqa 602` approach.

**Comment by aniro** [comment]
@geerlingguy 
The following: 
> ```
> jenkins_repo_url | default(false)
> ```
> 
> That filter seems to evaluate correctly to `False` if the URL is string var is empty, and doesn't give the bare conditional warning.

Does not work the other way round with default(true). So in your mysql role:
```mysql_log == ""```
is not equivalent to
```mysql_log | default(true)```
since the latter evaluates to false if the string is empty. 

Essentially there is no difference between providing `true` or `false` to the `default` filter:
```
$ cat test.yml
- hosts: localhost
  connection: local
  gather_facts: false

  vars:
    string_empty: ''
    string_non_empty: 'some string //'

  tasks:
    - name: string_empty | default(true)
      command: date
      when:
        - string_empty | default(true)

    - name: string_non_empty | default(true)
      command: date
      when:
        - string_non_empty | default(true)

    - name: string_empty | default(false)
      command: date
      when:
        - string_empty | default(false)

    - name: string_non_empty | default(false)
      command: date
      when:
        - string_non_empty | default(false)

$ ansible-playbook test.yml
PLAY [localhost] *******************************************************************************************************
TASK [string_empty | default(true)] ************************************************************************************
skipping: [localhost]

TASK [string_non_empty | default(true)] ********************************************************************************
changed: [localhost]

TASK [string_empty | default(false)] ***********************************************************************************
skipping: [localhost]

TASK [string_non_empty | default(false)] *******************************************************************************
changed: [localhost]```

Hence, using `string | default(true/false)` only worsens readability and confuses.
using `not string | default(false)` to check on empty string is even worse.
IMO string comparison is the only viable solution at the moment.

All above was tested on ansible 2.9.4

**Comment by cipy** [comment]
why not introducing a simple filter?

when: something|empty or not something|empty

🤔

**Comment by tgadiev** [comment]
What's the problem exactly? Simple `when: myvar` and `when: not myvar` now works as expected. Empty strings are treated as false, non-empty as true.

**Comment by cipy** [comment]
@tgadiev it is explained up in the thread

when: not myvar

is true when myvar is 'False', '0', and also '' (empty string)

**Comment by tgadiev** [comment]
@cipy handling undefined variables is a different case. I don't think it's related to this discussion about conditionals.

**Comment by jhanssenfavaro** [comment]
Well, normally needs to use both 'variable is defined and variable' that sounds weird!!!

**Comment by elafontaine** [comment]
Folks, I just got this issue and the default python evaluation was what I was expecting.  Now, I'm lost in regards of what is it I'm supposed to put.  I ran this on python 2 with ansible 2.7.11 and python3.6 with ansible 2.7.11. The surprise to me was that it did blow out with Python3 but not python2.

also, "|bool" doesn't work better on ansible 2.7 (as listed above).  

I'm testing this at the moment, but would like confirmation that this is the recommended way now.  (The role I'm making need to run for ansible 2.4 through 2.8) ;
```
  when:
  - var1 is defined
  - var1 | length > 0
```

**Comment by sjpb** [comment]
> Ah, asking around, I discovered that the `| bool` filter only returns `True` for _truthy_ string values (so like `'true'`, or `'True'` or `1`), but not for _non-empty strings_. Big miss on my part.

Except its not really a big miss seeing as this seems to be totally undocumented!

**Comment by ssbarnea** [comment]
@geerlingguy Do you think we have any chance to come to some conclusions on this famous ticket? I would really want to close it and while removal of the rule is a quick way to achieve it, I am bit worried that this rule prevented introduction of some bugs in the past.

**Comment by geerlingguy** [comment]
In my own work, I can't think of a time when this rule ever helped prevent a mistake. It only annoyed me because there are many times where a comparison to empty string is the _only possible_ way to do it right because of Python's duck typing.

I just add it to the skip list on all my playbooks now :(

**Comment by nixfu** [comment]
Why is this so controversial?  

when: MyString 
when: not MyString

This is functionally the same as the very common python usage of
if MyString:
if not MyString:


See this how to do it right(the way python does it): https://docs.python.org/3/library/stdtypes.html#truth-value-testing

**Comment by phemmer** [comment]
> Why is this so controversial? This should be the way to be able to check if a value exists/is empty or not.
> 
> when: MyString
> when: not MyString
> 
> This is functionally the same as the very common python usage of
> if MyString:
> if not MyString:
> 
> See this how to do it right(the way python does it): https://docs.python.org/3/library/stdtypes.html#truth-value-testing

See https://github.com/ansible-community/ansible-lint/issues/457#issuecomment-594118359

**Comment by elafontaine** [comment]
Maybe we could get a filter to make things "truthy" or "falsy" like python does... this way, we could abstract that problem away and get people to be explicit. 

Something like 
```
when: not truthy(string)
```

**Comment by cognifloyd** [comment]
Does anyone have examples where the `empty-string-compare` rule prevents bugs?

I asked in chat to understand why comparing with an empty string is dangerous, as described in @ssbarnea's [comment above](https://github.com/ansible-community/ansible-lint/issues/457#issuecomment-782668372):
> I am bit worried that this rule prevented introduction of some bugs in the past.

@ssbarnea said:
> sadly there is no solution fits all for all string compare
> I will have to reopen the subject at some point and document all options with pros and cons because last discussions were long time ago and I forgot some of the nuances.

So, can we create such a document together? What are the pros/cons of various string comparison alternatives? What bugs are possible (or common) with each alternative?

**Comment by IronTooch** [comment]
My two cents on the topic is that this rule should be removed. It makes an assumption that the user is an experienced Python user who ***knows*** that `mystring=""` evaluates to `false`. One of the major benefits of Ansible, I feel like, is that it abstracts away the Python for the most part. So assuming that the user is aware of that feature of Python, I feel, is maybe overzealous. Particularly from people who work in super-strongly-typed languages. Those folks might (reasonably) make the assumption that `""` is true and declared appropriately, and undeclared or uninitialized variables **only** evaluate to `false`. At no point does `== ""` result in LESS code clarity. Arguably it's more declarative. I (personally) feel code clarity should be a major point of the linting solution.

**Comment by ssbarnea** [comment]
@awcrosby @sivel @bcoca Can you please help us make a decision regarding this confusing rule? 

While I do know for sure that it did catch some bugs in the past it also seems to give false recommendations in some cases.

Unless we can fix it by getting a recommended use pattern that is safe to use I would see it demoting it to an opt-in in upcoming v6.

**Comment by bcoca** [comment]
opt-in seems best, there are many cases you want to be specific and compare to empty string (also confirms 'type')
